File: iOS/GroceryApp/Copilot/ImagePrefetcher.swift
import Foundation
import CouchbaseLiteSwift

/// Warms the on-disk image cache once, right after the first sync lands.
///
/// Images are the one part of this app that is *not* offline-first. Documents and vectors arrive
/// through replication and live in Couchbase Lite, but product and golden-shelf photos are S3
/// URLs fetched on demand, so a screen that has never been scrolled has never downloaded its
/// pictures. That is invisible in normal use and fatal in a demo: the whole story is going
/// offline and staying useful, and reaching Copilot for the first time in airplane mode would
/// show empty frames — or worse, require switching the network back on mid-narrative.
///
/// So once sync settles, every image URL the app can currently reach is pulled down in the
/// background. Deliberately fire-and-forget: failures are ignored, nothing blocks the UI, and
/// anything already cached is skipped by `ImageDownloadService` itself.
enum ImagePrefetcher {

    /// Bounded concurrency. Serial would take minutes over a conference connection; unbounded
    /// would open ~130 sockets at once and starve the replicator that just finished.
    private static let maxConcurrent = 6

    private static var hasRun = false

    /// Collects every image URL in the local database and downloads it into the disk cache.
    ///
    /// Runs at most once per launch — the guard matters because `onInitialSyncComplete` can fire
    /// again for a second endpoint, and re-walking the collections each time would be wasted
    /// work for no new images.
    static func warmCache(databaseManager: DatabaseManager) {
        guard !hasRun else { return }
        hasRun = true

        Task.detached(priority: .utility) {
            let urls = await collectURLs(databaseManager: databaseManager)
            guard !urls.isEmpty else { return }
            print("🖼️ [Prefetch] warming \(urls.count) images for offline use")

            let started = Date()
            await withTaskGroup(of: Void.self) { group in
                var issued = 0
                for url in urls {
                    if issued >= maxConcurrent { await group.next() }
                    issued += 1
                    group.addTask { _ = await ImageDownloadService.shared.downloadImage(from: url) }
                }
            }
            let elapsed = Date().timeIntervalSince(started)
            print(String(format: "🖼️ [Prefetch] done in %.1fs", elapsed))
        }
    }

    /// Product images plus golden shelf references — the two sets the copilot renders.
    ///
    /// Read on a background context because this walks every inventory and planogram document,
    /// and it is triggered from a replicator callback that should not be held up.
    private static func collectURLs(databaseManager: DatabaseManager) async -> [String] {
        await Task.detached(priority: .utility) { () -> [String] in
            guard let database = await databaseManager.database else { return [] }
            var urls: [String] = []

            // `IS VALUED` filters out documents whose URL is missing or null, so the download
            // queue never contains entries that can only fail.
            let queries = [
                """
                SELECT imageURL FROM `\(AppConfig.scopeName)`.`\(AppConfig.collectionName)`
                WHERE imageURL IS VALUED
                """,
                """
                SELECT goldenImageURL AS imageURL
                FROM `\(AppConfig.scopeName)`.`\(AppConfig.planogramsCollectionName)`
                WHERE docType = "Planogram" AND goldenImageURL IS VALUED
                """
            ]
            for sql in queries {
                do {
                    for row in try database.createQuery(sql).execute() {
                        if let url = row.string(forKey: "imageURL"), !url.isEmpty {
                            urls.append(url)
                        }
                    }
                } catch {
                    print("⚠️ [Prefetch] URL collection failed: \(error)")
                }
            }
            // The same product image can appear on several documents; downloading it twice is
            // harmless but pointless.
            return Array(Set(urls))
        }.value
    }
}

File: Android/app/src/main/java/com/example/groceryapplication/copilot/ImagePrefetcher.kt
package com.example.groceryapplication.copilot

import android.content.Context
import android.util.Log
import coil.imageLoader
import coil.request.CachePolicy
import coil.request.ImageRequest
import com.couchbase.lite.Database
import com.example.groceryapplication.AppConfig
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch
import kotlinx.coroutines.sync.Semaphore
import kotlinx.coroutines.sync.withPermit
import kotlinx.coroutines.coroutineScope
import java.util.concurrent.atomic.AtomicBoolean

/**
 * Warms Coil's disk cache once, right after the first sync lands.
 *
 * Images are the one part of this app that is *not* offline-first. Documents and vectors arrive
 * through replication and live in Couchbase Lite, but product and golden-shelf photos are S3
 * URLs fetched on demand, so a screen that has never been scrolled has never downloaded its
 * pictures. That is invisible in normal use and fatal in a demo: the whole story is going
 * offline and staying useful, and reaching Copilot for the first time in airplane mode would
 * show empty frames — or worse, require switching the network back on mid-narrative.
 *
 * Requests go through the app's own Coil `imageLoader`, so what lands here is exactly what
 * `AsyncImage` reads later. Fire-and-forget: failures are ignored and nothing blocks the UI.
 *
 * Kotlin counterpart of iOS `ImagePrefetcher.swift`.
 */
object ImagePrefetcher {

    private const val TAG = "ImagePrefetch"

    /** Bounded concurrency: serial is too slow over conference wifi, unbounded would open ~130
     *  sockets at once and starve the replicator that just finished. */
    private const val MAX_CONCURRENT = 6

    private val hasRun = AtomicBoolean(false)
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)

    /**
     * Collects every image URL in the local database and pulls it into the disk cache.
     *
     * Runs at most once per launch — `onInitialSyncComplete` can fire again for a second
     * endpoint, and re-walking the collections would be wasted work for no new images.
     */
    fun warmCache(context: Context, database: Database?) {
        if (database == null) return
        if (!hasRun.compareAndSet(false, true)) return

        val appContext = context.applicationContext
        scope.launch {
            val urls = collectUrls(database)
            if (urls.isEmpty()) return@launch
            Log.i(TAG, "🖼️ warming ${urls.size} images for offline use")

            val started = System.currentTimeMillis()
            val gate = Semaphore(MAX_CONCURRENT)
            coroutineScope {
                urls.forEach { url ->
                    launch {
                        gate.withPermit {
                            val request = ImageRequest.Builder(appContext)
                                .data(url)
                                // Write to disk, skip the memory cache: this is about surviving
                                // a later cold render offline, not about being resident now.
                                .diskCachePolicy(CachePolicy.ENABLED)
                                .memoryCachePolicy(CachePolicy.DISABLED)
                                .build()
                            runCatching { appContext.imageLoader.execute(request) }
                        }
                    }
                }
            }
            Log.i(TAG, "🖼️ done in ${System.currentTimeMillis() - started} ms")
        }
    }

    /** Product images plus golden shelf references — the two sets the copilot renders. */
    private fun collectUrls(database: Database): List<String> {
        val urls = mutableListOf<String>()
        // `IS VALUED` filters out documents whose URL is missing or null, so the download queue
        // never contains entries that can only fail.
        val queries = listOf(
            """
            SELECT imageURL FROM `${AppConfig.scopeName}`.`${AppConfig.COLLECTION_NAME}`
            WHERE imageURL IS VALUED
            """.trimIndent(),
            """
            SELECT goldenImageURL AS imageURL
            FROM `${AppConfig.scopeName}`.`${AppConfig.PLANOGRAMS_COLLECTION_NAME}`
            WHERE docType = "Planogram" AND goldenImageURL IS VALUED
            """.trimIndent()
        )
        queries.forEach { sql ->
            try {
                database.createQuery(sql).execute().forEach { row ->
                    row.getString("imageURL")?.takeIf { it.isNotEmpty() }?.let { urls += it }
                }
            } catch (t: Throwable) {
                Log.w(TAG, "⚠️ URL collection failed", t)
            }
        }
        // The same product image can appear on several documents; downloading it twice is
        // harmless but pointless.
        return urls.distinct()
    }
}
